### PR TITLE
Stock management page

### PR DESCRIPTION
**Pull request by nihco2**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add stock management page
| Type?         | new feature
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | (optional)
| How to test?  | Refers to prestaforge FF board

**TODO:** 
- [x] Script to update ps_tab_lang

### CODE REVIEW

File: src/PrestaShopBundle/Controller/Api/StockController.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<?php
+/**
+ * 2007-2016 PrestaShop
```
Comment (aleeks) [inline src/PrestaShopBundle/Controller/Api/StockController.php]: No more 2016 please !!!

Comment (nihco2) [inline src/PrestaShopBundle/Controller/Api/StockController.php]: yes !

File: admin-dev/themes/new-theme/js/stock-page/components/app.vue [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<template>
```
Comment (kpodemski) [inline admin-dev/themes/new-theme/js/stock-page/components/app.vue]: HELL YEAH! 👍 Vue in PrestaShop :D

File: admin-dev/themes/new-theme/js/stock-page/components/header/search-filter.vue [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<template>
+  <select class="form-control m-b-3 search-filter">
+    <option selected>Open this select menu</option>
```
Comment (aleeks) [inline admin-dev/themes/new-theme/js/stock-page/components/header/search-filter.vue]: Yooooooooooo!
Where do the translations will come from ?!!

Comment (xBorderie) [inline admin-dev/themes/new-theme/js/stock-page/components/header/search-filter.vue]: ?!!... :p

Comment (aleeks) [inline admin-dev/themes/new-theme/js/stock-page/components/header/search-filter.vue]: :smile_cat:

Comment (nihco2) [inline admin-dev/themes/new-theme/js/stock-page/components/header/search-filter.vue]: @aleeks We will do that in the next step, probably with a json generated on server side. ping @Shudrum @thierrymarianne

File: install-dev/upgrade/sql/1.7.2.0.sql [hunk 4]
```diff
@@ -0,0 +1,3 @@
+
+ALTER TABLE  `PREFIX_stock_available` ADD  `physical_quantity` INT NOT NULL DEFAULT  '0' AFTER  `quantity`;
+ALTER TABLE  `PREFIX_stock_available` ADD  `reserved_quantity` INT NOT NULL DEFAULT  '0' AFTER  `quantity`;
```
Comment (aleeks) [inline install-dev/upgrade/sql/1.7.2.0.sql]: @thierrymarianne 
I think you have to add an entry for AdminStockManagement tab! (like your xml/tab.xml)

Comment (aleeks) [inline install-dev/upgrade/sql/1.7.2.0.sql]: @thierrymarianne After reserved_quantity (to be equal with db_structure) ?

File: robots.txt [hunk 5]
```diff
@@ -78,18 +78,18 @@ Disallow: */vendor/
...
+Disallow: /discount
+Disallow: /order-history
+Disallow: /identity
+Disallow: /my-account
+Disallow: /order-follow
+Disallow: /credit-slip
+Disallow: /order
+Disallow: /search
+Disallow: /guest-tracking
+Disallow: /order-confirmation
```
Comment (aleeks) [inline robots.txt]: Maybe disappear after a rebase!

File: admin-dev/themes/new-theme/js/stock-page/components/products/alert.vue [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      <div class="alert alert-info" role="alert">
+        <i class="material-icons">info_outline</i>
+        <p class="alert-text">No product matches your search. Try changing search terms.</p>
+      </div>
+    </td>
+  </tr>
+</template>
+
+<style lang="sass?outputStyle=expanded" scoped>
+  @import "~PrestaKit/scss/custom/_variables.scss";
```
Comment (mickaelandrieu) [inline admin-dev/themes/new-theme/js/stock-page/components/products/alert.vue]: if you use "scoped", import the variables on the Main app/component: you are duplicating the file in a unique generated id.

File: admin-dev/themes/new-theme/js/stock-page/components/products/product/spinner.vue [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        }
+        this.isEnabled = !!this.value;
+      },
+      sendQty(event) {
+        let apiRootUrl = data.apiRootUrl.replace(/\?.*/,'');
+        let apiEditProductsUrl = `${apiRootUrl}/product/${this.product.product_id}`
+        let apiEditCombinationsUrl = `${apiRootUrl}/product/${this.product.product_id}/combination/${this.product.combination_id}`;
+        let postUrl = this.product.combination_id ? apiEditCombinationsUrl : apiEditProductsUrl;
+
+        event.preventDefault();
```
Comment (mickaelandrieu) [inline admin-dev/themes/new-theme/js/stock-page/components/products/product/spinner.vue]: you never should do that in Vue, take a look at https://vuejs.org/v2/guide/events.html#Event-Modifiers

File: admin-dev/themes/new-theme/js/stock-page/components/products/alert.vue (hunk 6 above)
Comment (nihco2) [inline admin-dev/themes/new-theme/js/stock-page/components/products/alert.vue]: Hey @mickaelandrieu, thx for your review

File: admin-dev/themes/new-theme/js/stock-page/store/index.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+const state = {
+  order: 'product',
+  pageIndex: 0,
+  totalPages: 0,
+  productsPerPage: 100,
+  combinationsPerPage: 50,
+  keywords: []
+};
+
+// getters are functions
```
Comment (mickaelandrieu) [inline admin-dev/themes/new-theme/js/stock-page/store/index.js]: if getters are functions, why use "const" operators?

File: admin-dev/themes/new-theme/js/stock-page/store/modules/products.js [hunk 9]
```diff
@@ -0,0 +1,117 @@
+import * as types from '../mutation-types';
+import Vue from 'vue';
+import VueResource from 'vue-resource';
+import { showGrowl } from '../utils/growl';
+
+Vue.use(VueResource);
+
+// initial state
+const state = {
```
Comment (mickaelandrieu) [inline admin-dev/themes/new-theme/js/stock-page/store/modules/products.js]: the right behavior of state is to never been able to change or be updated? You probably want to use `let` or `var`

File: admin-dev/themes/new-theme/js/stock-page/components/products/product/spinner.vue (hunk 7 above)
Comment (kpodemski) [inline admin-dev/themes/new-theme/js/stock-page/components/products/product/spinner.vue]: @mickaelandrieu oh, i think that you can use it but you also can use it as a shorthand in a element:
`<form v-on:submit.prevent="onSubmit"></form>`

File: admin-dev/themes/new-theme/js/stock-page/store/modules/products.js (hunk 9 above)
Comment (nihco2) [inline admin-dev/themes/new-theme/js/stock-page/store/modules/products.js]: They use const in Vuex doc, I don't know what is the best practice

File: admin-dev/themes/new-theme/package.json [hunk 10]
```diff
@@ -4,47 +4,55 @@
...
     "sass-loader": "^3.1.2",
     "sprintf-js": "^1.0.3",
     "style-loader": "^0.13.0",
     "tether": "^1.3.2",
     "typeahead.js": "^0.11.1",
-    "webpack": "^1.12.12"
-  },
-  "dependencies": {
-    "mobile-detect": "^1.3.5"
+    "vue": "^2.1.10",
```
Comment (mickaelandrieu) [inline admin-dev/themes/new-theme/package.json]: :'( it's a great victory for PrestaShop back office, thank you to have started this!

File: src/Adapter/ImageManager.php [hunk 11]
```diff
@@ -56,18 +56,80 @@ public function __construct(LegacyContext $legacyContext)
...
         // because legacy uses relative path to reach a directory under root directory...
         $replacement = 'src="'.$this->legacyContext->getRootUrl();
         $thumbPath = preg_replace('/src="(\\.\\.\\/)+/', $replacement, $thumbPath);
 
         return $thumbPath;
     }
+
+    public function getThumbnailPath($imageId)
+    {
+        $imageType = 'jpg';
```
Comment (mickaelandrieu) [inline src/Adapter/ImageManager.php]: should be constants

File: admin-dev/themes/new-theme/js/stock-page/store/modules/products.js (hunk 9 above)
Comment (kpodemski) [inline admin-dev/themes/new-theme/js/stock-page/store/modules/products.js]: @mickaelandrieu yep, `const` is not immutable, but it's js you know how it is :D

File: src/Adapter/ImageManager.php (hunk 11 above)
Comment (nihco2) [inline src/Adapter/ImageManager.php]: ping @aleeks

File: admin-dev/themes/new-theme/package.json (hunk 10 above)
Comment (xBorderie) [inline admin-dev/themes/new-theme/package.json]: Why the sad face?

File: src/PrestaShopBundle/Controller/Admin/StockController.php [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+     * @return array Template vars
+     */
+    public function overviewAction()
+    {
+        return [];
+    }
+
+    public function hashUpdateJsAction($hash)
+    {
+        $contents = file_get_contents('http://localhost:8080/' . $hash . '.hot-update.js');
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Admin/StockController.php]: it's a wip right?

File: admin-dev/themes/new-theme/package.json (hunk 10 above)
Comment (kpodemski) [inline admin-dev/themes/new-theme/package.json]: @xBorderie it's like: "damn, i left and they started to use cool stuff" :D i hope!

File: src/PrestaShopBundle/Controller/Api/StockController.php [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+
+    /**
+     * @param Request $request
+     * @return JsonResponse
+     */
+    public function editProductAction(Request $request)
+    {
+        try {
+            $this->guardAgainstMissingDeltaParameter($request);
+            $delta = (int)$request->request->get('delta');
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Controller/Api/StockController.php]:
```diff
- $delta = (int)$request->request->get('delta');
+ $delta = $request->request->getInt('delta');
```

File: src/PrestaShopBundle/Controller/Admin/StockController.php (hunk 12 above)
Comment (nihco2) [inline src/PrestaShopBundle/Controller/Admin/StockController.php]: yep

File: admin-dev/themes/new-theme/js/stock-page/components/products/alert.vue (hunk 6 above)
Comment (nihco2) [inline admin-dev/themes/new-theme/js/stock-page/components/products/alert.vue]: Same problem as https://github.com/vuejs/vue-loader/issues/328
I will try to resolve that with sass loader in webpack

**Comment by Dh42** [comment]
https://i.imgur.com/NkDJuLQ.jpg

**Comment by aleeks** [comment]
Hi, translations PR have these commits

**Pull request closed without merging**